Saule/Serialization/ResourceGraph.cs

```diff
@@ -0,0 +1,144 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+using Saule.Queries.Including;
+
+namespace Saule.Serialization
+{
+    internal class ResourceGraph
+    {
+        private readonly IDictionary<ResourceGraphNodeKey, ResourceGraphNode> _nodes;
+
+        public ResourceGraph(
+            object obj,
+            ApiResource resource,
+            ResourceGraphPathSet includePaths)
+        {
+            obj.ThrowIfNull(nameof(obj));
+            resource.ThrowIfNull(nameof(resource));
+            includePaths.ThrowIfNull(nameof(includePaths));
+
+            _nodes = new Dictionary<ResourceGraphNodeKey, ResourceGraphNode>();
+
+            Build(obj, resource, includePaths, 0);
+        }
+
+        public IEnumerable<ResourceGraphNode> DataNodes
+        {
+            get
+            {
+                return _nodes.Values.Where(v => v.GraphDepth == 0);
+            }
+        }
+
+        public IEnumerable<ResourceGraphNode> IncludedNodes
+        {
+            get
+            {
+                return _nodes.Values.Where(v => v.GraphDepth > 0);
+            }
+        }
+
+        private void Build(
+            object obj,
+            ApiResource resource,
+            ResourceGraphPathSet includePaths,
+            int depth)
+        {
+            if (obj == null)
+            {
+                // end of the line - obj will be a leaf node in our graph
+                return;
+            }
+
+            if (obj is IEnumerable)
```

Comment:
I was half-referring to #99. If I have an object graph like this, where every property is lazy (say, using an ORM):

```
- Person
  - Job
    - LiterallyEveryEmailEverSent     // huge!
```

but no Saule resource ever refers to `LiterallyEveryEmailEverSent`:

```csharp
class PersonResource : ApiResource
{
  public PersonResource() {
    BelongsTo<CompanyResource>(nameof(Person.Job));
  }
}

class CompanyResource : ApiResource
{
  public CompanyResource() {
  }
}
```

The current serializer will materialize `LiterallyEveryEmailEverSent`, which is obviously wasteful because it won't end up in the final JSON output. Does the new serializer still have this effect, or is it circumvented somehow? It's not immediately obvious to me from the code (these things can be tricky).

Since the current serializer *does* materialize everything, I think it's OK if the new one does too; we can always improve on it later. That said, if it already *doesn't* do it, that would be better :smile: In that case, I think tests are necessary, because it seems pretty easy to mess up somewhere and evaluate everything when someone changes something later.